File: flask_app/app/dao/discount_usage_dao.py
from app.models.discount_usage import DiscountUsage
from app.models import db

class DiscountUsageDAO:

    @staticmethod
    def get_by_filter(page=1, per_page=10, filters=None):
        query = DiscountUsage.query
        filters = filters or {}

        if filters.get("user_id"):
            query = query.filter(DiscountUsage.user_id == filters["user_id"])

        if filters.get("discount_id"):
            query = query.filter(DiscountUsage.discount_id == filters["discount_id"])

        pagination = query.paginate(page=page, per_page=per_page, error_out=False)
        return pagination.items, pagination.total

    @staticmethod
    def get_all(page=1, per_page=10, filters=None):
        return DiscountUsageDAO.get_by_filter(page=page, per_page=per_page, filters=filters)

    @staticmethod
    def get_by_id(usage_id):
        return DiscountUsage.query.filter_by(id=usage_id).first()

    @staticmethod
    def get_by_user(user_id, page=1, per_page=10):
        return DiscountUsageDAO.get_by_filter(page=page, per_page=per_page,filters= {"user_id":user_id})
    
    @staticmethod
    def get_user_usage_count(user_id):
        return DiscountUsage.query.filter(DiscountUsage.user_id == user_id).count()

    @staticmethod
    def create(data):
        usage = DiscountUsage(**data)
        db.session.add(usage)
        db.session.commit()
        return usage

File: java_membership_api/src/main/java/com/example/membership_api/dao/DiscountUsageDAO.java
package com.example.membership_api.dao;

import java.util.List;

import org.springframework.stereotype.Repository;

import com.example.membership_api.model.DiscountUsage;

import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import jakarta.persistence.TypedQuery;
import jakarta.transaction.Transactional;

@Repository
@Transactional
public class DiscountUsageDAO {

    @PersistenceContext
    private EntityManager em;

    public List<DiscountUsage> getAll(int page, int perPage, String userId, String discountId) {
        StringBuilder queryString = new StringBuilder("FROM DiscountUsage du WHERE 1=1 ");

        if (userId != null) {
            queryString.append("AND du.userId = :userId ");
        }
        if (discountId != null) {
            queryString.append("AND du.discountId = :discountId ");
        }

        var query = em.createQuery(queryString.toString(), DiscountUsage.class);

        if (userId != null) {
            query.setParameter("userId", userId);
        }
        if (discountId != null) {
            query.setParameter("discountId", discountId);
        }

        query.setFirstResult((page - 1) * perPage);
        query.setMaxResults(perPage);

        return query.getResultList();
    }

    public long getTotal(String userId, String discountId) {
        StringBuilder queryString = new StringBuilder("SELECT COUNT(du) FROM DiscountUsage du WHERE 1=1 ");

        if (userId != null) {
            queryString.append("AND du.userId = :userId ");
        }
        if (discountId != null) {
            queryString.append("AND du.discountId = :discountId ");
        }

        var query = em.createQuery(queryString.toString(), Long.class);

        if (userId != null) {
            query.setParameter("userId", userId);
        }
        if (discountId != null) {
            query.setParameter("discountId", discountId);
        }

        return query.getSingleResult();
    }

    public DiscountUsage getById(String id) {
        return em.find(DiscountUsage.class, id);
    }

    public List<DiscountUsage> getByUser(String userId, int page, int perPage) {
        TypedQuery<DiscountUsage> query = em.createQuery(
                "FROM DiscountUsage du WHERE du.userId = :userId", DiscountUsage.class);
        query.setParameter("userId", userId);
        query.setFirstResult((page - 1) * perPage);
        query.setMaxResults(perPage);
        return query.getResultList();
    }

    public long getUserUsageCount(String userId) {
        TypedQuery<Long> query = em.createQuery(
                "SELECT COUNT(du) FROM DiscountUsage du WHERE du.userId = :userId", Long.class);
        query.setParameter("userId", userId);
        return query.getSingleResult();
    }

    public DiscountUsage create(DiscountUsage usage) {
        em.persist(usage);
        return usage;
    }

    public DiscountUsage update(DiscountUsage existing, DiscountUsage updates) {
        existing.setDiscountId(updates.getDiscountId());
        existing.setUserId(updates.getUserId());
        existing.setAppliedOn(updates.getAppliedOn());
        existing.setEntity(updates.getEntity());
        existing.setDiscountAmount(updates.getDiscountAmount());
        existing.setUsedAt(updates.getUsedAt());
        return existing;
    }

    public boolean softDelete(String id) {
        DiscountUsage usage = getById(id);
        if (usage == null) return false;
        // You might want to add an "active" flag on DiscountUsage for soft delete
        // If not, this method can simply return false or throw exception.
        return false; // Or implement soft delete logic here
    }
}
